pkg/sdn/plugin/sdn-cni-plugin/openshift-sdn.go

```diff
@@ -0,0 +1,463 @@
+		}
+		if len(addrs) == 0 {
+			return fmt.Errorf("container had no addresses")
+		}
+		podIP = addrs[0].IP.String()
+
+		return nil
+	})
+	if err != nil {
+		return nil, "", "", fmt.Errorf("Failed to inspect container interface: %v", err)
+	}
+
+	hostVeth, err := netlink.LinkByIndex(peerIfindex)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("Failed to get host veth: %v", err)
+	}
+
+	return hostVeth, contVeth.Attrs().HardwareAddr.String(), podIP, nil
+}
+
+func addMacvlan(netns string) error {
+	var defIface netlink.Link
+	var err error
+
+	// Find interface with the default route
+	routes, _ := netlink.RouteList(nil, netlink.FAMILY_V4)
+	for _, r := range routes {
+		if r.Dst == nil {
+			defIface, err = netlink.LinkByIndex(r.LinkIndex)
+			if err != nil {
+				return fmt.Errorf("Failed to get default route interface: %v", err)
+			}
+		}
+	}
+	if defIface == nil {
+		return fmt.Errorf("Failed to find default route interface")
+	}
+
+	containerNs, err := ns.GetNS(netns)
+	if err != nil {
+		return fmt.Errorf("Failed to get container netns: %v", err)
+	}
+	defer containerNs.Close()
+
+	return containerNs.Do(func(ns.NetNS) error {
+		err := netlink.LinkAdd(&netlink.Macvlan{
+			LinkAttrs: netlink.LinkAttrs{
+				MTU:         defIface.Attrs().MTU,
+				Name:        "macvlan0",
+				ParentIndex: defIface.Attrs().Index,
+			},
+			Mode: netlink.MACVLAN_MODE_PRIVATE,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create macvlan interface: %v", err)
+		}
+		l, err := netlink.LinkByName("macvlan0")
+		if err != nil {
+			return fmt.Errorf("failed to create find macvlan interface: %v", err)
+		}
+		err = netlink.LinkSetUp(l)
+		if err != nil {
+			return fmt.Errorf("failed to set macvlan interface up: %v", err)
+		}
+		return nil
+	})
+}
+
+func getIPAMConfig(nodeConfig *NodeConfig) []byte {
+	_, ipnet, _ := net.ParseCIDR(nodeConfig.NodeNetwork)
+	nodeGateway := netutils.GenerateDefaultGateway(ipnet).String()
+	return []byte(fmt.Sprintf(ipamConfigTemplate, nodeConfig.NodeNetwork, nodeGateway, nodeConfig.ClusterNetwork))
+}
+
+func isScriptError(err error) bool {
+	_, ok := err.(*exec.ExitError)
+	return ok
+}
+
+// Get the last command (which is prefixed with "+" because of "set -x") and its output
+func getScriptError(output []byte) string {
+	lines := strings.Split(string(output), "\n")
+	for n := len(lines) - 1; n >= 0; n-- {
+		if strings.HasPrefix(lines[n], "+") {
+			return strings.Join(lines[n:], "\n")
+		}
+	}
+	return string(output)
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	nodeConfig, vnid, ingress, egress, macvlan, err := getPodInfo(args)
+	if err != nil {
+		return err
+	}
+
+	// Run IPAM so we can set up container veth
+	ipamConfig := getIPAMConfig(nodeConfig)
+	os.Setenv("CNI_ARGS", "")
+	result, err := ipam.ExecAdd("host-local", ipamConfig)
+	if err != nil {
+		return fmt.Errorf("Failed to run CNI IPAM ADD: %v", err)
+	}
+	if result.IP4 == nil || result.IP4.IP.IP.To4() == nil {
```

Comment:
Yep, result.IP4 is a structure of address, DNS, and routes.  result.IP4.IP is a net.IPNet.  And we want to ensure that the IP address (not IPNet) we're getting is an IPv4 address, thus result.IP4.IP.IP.To4().